fvm/evm/types/block.go

```diff
@@ -217,3 +239,70 @@ func (t TransactionHashes) EncodeIndex(index int, buffer *bytes.Buffer) {
 func (t TransactionHashes) RootHash() gethCommon.Hash {
 	return gethTypes.DeriveSha(t, gethTrie.NewStackTrie(nil))
 }
+
+// Below block type section, defines earlier block types,
+// this is being used to decode blocks that were stored
+// before block type changes. It allows us to still decode
+// a block that would otherwise be invalid if decoded into
+// latest version of the above Block type.
+
+// before adding PrevRandao to the block
+type BlockV0 struct {
+	ParentBlockHash     gethCommon.Hash
+	Height              uint64
+	Timestamp           uint64
+	TotalSupply         *big.Int
+	ReceiptRoot         gethCommon.Hash
+	TransactionHashRoot gethCommon.Hash
+	TotalGasUsed        uint64
+}
+
+type BlockProposalV0 struct {
+	BlockV0
+	Receipts []LightReceipt
+	TxHashes TransactionHashes
+}
+
+// decodeBlockBreakingChanges will try to decode the bytes into all
+// previous versions of block type, if it succeeds it will return the
+// migrated block, otherwise it will return nil.
+func decodeBlockBreakingChanges(encoded []byte) *Block {
+	b0 := &BlockV0{}
+	if err := gethRLP.DecodeBytes(encoded, b0); err == nil {
+		return &Block{
+			ParentBlockHash:     b0.ParentBlockHash,
+			Height:              b0.Height,
+			Timestamp:           b0.Timestamp,
+			TotalSupply:         b0.TotalSupply,
+			ReceiptRoot:         b0.ReceiptRoot,
+			TransactionHashRoot: b0.TransactionHashRoot,
+			TotalGasUsed:        b0.TotalGasUsed,
+			PrevRandao:          gethCommon.Hash{},
+		}
+	}
+	return nil
+}
+
+// decodeBlockProposalBreakingChanges will try to decode the bytes into all
+// previous versions of block proposal type, if it succeeds it will return the
+// migrated block, otherwise it will return nil.
```

Comment:
Maybe it should return an error on failure...

Reply:
when we return the nil, on the caller we have a nil check return the original (non-legacy decoding) error which would be wrapped as a failure.